src/storage/src/types/errors.rs

```diff
@@ -50,22 +50,200 @@ impl Codec for DecodeError {
+            }),
+        }
+    }
+
+    fn from_proto(proto: ProtoDecodeErrorInner) -> Result<Self, TryFromProtoError> {
+        use proto_decode_error_inner::Kind::*;
+        match proto.kind {
+            Some(Text(v)) => Ok(DecodeErrorInner::Text(v)),
+            None => Err(TryFromProtoError::missing_field("ProtoDecodeError::kind")),
+        }
+    }
+}
+
+impl Display for DecodeErrorInner {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            DecodeErrorInner::Text(e) => write!(f, "Text: {}", e),
+        }
+    }
+}
+
+/// Errors arising during envelope processing.
+#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
+pub enum EnvelopeError {
+    /// An error arising while processing the Debezium envelope.
+    Debezium(String),
+    /// An error that can be retracted by a future message using upsert logic.
+    Upsert(UpsertError),
+    /// Errors corresponding to `ENVELOPE NONE`. Naming this
+    /// `None`, though, would have been too confusing.
+    Flat(String),
+}
+
+impl RustType<ProtoEnvelopeError> for EnvelopeError {
+    fn into_proto(&self) -> ProtoEnvelopeError {
+        use proto_envelope_error::Kind;
+        ProtoEnvelopeError {
+            kind: Some(match self {
+                EnvelopeError::Debezium(text) => Kind::Debezium(text.clone()),
+                EnvelopeError::Upsert(rust) => Kind::Upsert(Box::new(rust.into_proto())),
+                EnvelopeError::Flat(text) => Kind::Flat(text.clone()),
+            }),
+        }
+    }
+
+    fn from_proto(proto: ProtoEnvelopeError) -> Result<Self, TryFromProtoError> {
+        use proto_envelope_error::Kind;
+        match proto.kind {
+            Some(Kind::Debezium(text)) => Ok(Self::Debezium(text)),
+            Some(Kind::Upsert(proto)) => {
+                let rust = RustType::from_proto(*proto)?;
+                Ok(Self::Upsert(rust))
+            }
+            Some(Kind::Flat(text)) => Ok(Self::Flat(text)),
+            None => Err(TryFromProtoError::missing_field("ProtoEnvelopeError::kind")),
+        }
+    }
+}
+
+impl Display for EnvelopeError {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
-            DecodeError::Text(e) => write!(f, "Text: {}", e),
+            EnvelopeError::Debezium(err) => write!(f, "Debezium: {err}"),
+            EnvelopeError::Upsert(err) => write!(f, "Upsert: {err}"),
+            EnvelopeError::Flat(err) => write!(f, "Flat: {err}"),
+        }
+    }
+}
+
+/// An error from a value in an upsert source. The corresponding key is included, allowing
+/// us to reconstruct their entry in the upsert map upon restart.
+#[derive(Ord, PartialOrd, Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash)]
+pub struct UpsertValueError {
+    /// The underlying error.
+    pub inner: Box<DataflowError>,
+    /// The (good) key associated with the errored value.
+    pub for_key: Row,
```

Comment:
do we need this here? don't we have that key already in the `current_values` map?

Reply:
Yes; we don't have it in the persist shard, though! The persist shard only holds the values, not the `(k, v)` pair. That's why we needed to add this data -- so we can reconstruct the `(k, v)` pair from just the error.

Reply:
oh yeah, i forgot about that quirk